feat: map additional fields from mobile agents

kruskall opened the pull request:
<!-- Thanks for sending a pull request!

If this is your first contribution, please review and sign our contributor agreement -
https://www.elastic.co/contributor-agreement.

Guidelines:
 - Prefer small PRs, and split changes into multiple logical commits where they must
   be delivered in a single PR.
 - If the PR is incomplete and not yet ready for review, open it as a Draft.
 - Once the PR is marked ready for review it is expected to pass all tests and linting,
   and you should not force-push any changes.

See also https://github.com/elastic/apm-server/blob/main/CONTRIBUTING.md for more tips on contributing.
-->

## Motivation/summary

Add mapping for 'host.os.version' and 'host.os.name'.
Android and ios are recognized and os type is set
accordingly.

## Checklist

<!--
Delete irrelevant items. The changelog should only be updated for user-facing changes.
Once the PR is ready for review there should be no unticked boxes.
-->

- [x] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/main/CHANGELOG.asciidoc)
- [ ] Update [package changelog.yml](https://github.com/elastic/apm-server/blob/main/apmpackage/apm/changelog.yml) (only if changes to `apmpackage` have been made)
- [ ] Documentation has been updated

For functional changes, consider:
- Is it observable through the addition of either **logging** or **metrics**?
- Is its use being published in **telemetry** to enable product improvement?
- Have system tests been added to avoid regression?

## How to test these changes

<!--
Explain how this PR can be tested by the reviewer: commands, dependencies, steps, etc.
If it is self-explanatory, delete this section.
-->

## Related issues

Closes https://github.com/elastic/apm-server/issues/8755

apmmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-8784/detail/PR-8784/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-8784/detail/PR-8784/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-8784/detail/PR-8784/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-8784/detail/PR-8784/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://apm-server_8784.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/apm-ci/transactions/view?rangeFrom=2022-08-02T20:44:14.938Z&rangeTo=2022-08-02T21:04:14.938Z&transactionName=BUILD+apm-server%2Fapm-server-mbp%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=4bca6e90c6eee6c8358082e1f4487ee0&transactionId=4766841f1759ac48)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2022-08-02T20:54:14.938+0000


* Duration: 6 min 26 sec





</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->














### Steps errors [![1](https://img.shields.io/badge/1%20-red)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-8784/detail/PR-8784/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `Run intake`
<ul>
<li>Took 3 min 55 sec . View more details <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-server/pipelines/apm-server-mbp/pipelines/PR-8784/runs/1/steps/158/log/?start=0">here</a></li>
<li>Description: <code>./.ci/scripts/intake.sh</code></l1>
</ul>
  
  </p>
  </details>



## :robot: GitHub comments

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `/package` : Generate and publish the docker images.


- `/test windows` : Build & tests on Windows.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




<!--COMMENT_GENERATED_WITH_ID_comment.id-->

apmmachine commented:
### :books: Go benchmark report

Diff with the `main` branch

```
name                                                            old time/op    new time/op    delta
pkg:github.com/elastic/apm-server/internal/agentcfg goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/internal/decoder goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/internal/model/modelindexer goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/internal/processor/stream goos:linux goarch:amd64
BackendProcessorParallel/BenchmarkBackendProcessorParallel2-12    0.00ns ±32%    0.00ns ±22%  -44.21%  (p=0.008 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel4-12    0.00ns ±15%    0.00ns ±48%  -44.15%  (p=0.008 n=5+5)
pkg:github.com/elastic/apm-server/internal/publish goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/aggregation/spanmetrics goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/aggregation/txmetrics goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/sampling goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/sampling/eventstorage goos:linux goarch:amd64
IsTraceSampled/unsampled-12                                       77.5ns ± 1%    79.2ns ± 2%   +2.23%  (p=0.008 n=5+5)

name                                                            old alloc/op   new alloc/op   delta
pkg:github.com/elastic/apm-server/internal/agentcfg goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/internal/decoder goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/internal/model/modelindexer goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/internal/processor/stream goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/internal/publish goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/aggregation/spanmetrics goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/aggregation/txmetrics goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/sampling goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/sampling/eventstorage goos:linux goarch:amd64

name                                                            old allocs/op  new allocs/op  delta
pkg:github.com/elastic/apm-server/internal/agentcfg goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/internal/decoder goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/internal/model/modelindexer goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/internal/processor/stream goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/internal/publish goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/aggregation/spanmetrics goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/aggregation/txmetrics goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/sampling goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/sampling/eventstorage goos:linux goarch:amd64

```

_report generated with https://pkg.go.dev/golang.org/x/perf/cmd/benchstat_
<!--COMMENT_GENERATED_WITH_ID_bench.diff-->

axw commented on internal/processor/otel/metadata.go:
```diff
@@ -153,6 +157,13 @@ func translateResourceMetadata(resource pcommon.Resource, out *model.APMEvent) {
 		out.Host.OS.Type = "unix"
 	}
 
+	switch out.Host.Name {
```
```suggestion
	switch out.Host.OS.Name {
```

kruskall replied:
Good catch! Changed 👍

axw commented on internal/processor/otel/metadata.go:
```diff
@@ -153,6 +157,13 @@ func translateResourceMetadata(resource pcommon.Resource, out *model.APMEvent) {
 
+	switch out.Host.Name {
+	case "Android":
+		out.Host.Type = "android"
+	case "iOS":
+		out.Host.Type = "ios"
```
```suggestion
		out.Host.OS.Type = "android"
	case "iOS":
		out.Host.OS.Type = "ios"
```

axw reviewed (commented):
Can you please add tests which show that `host.os.{name,type}` are set correctly for iOS and Android?

Also, these changes aren't quite enough to close https://github.com/elastic/apm-server/issues/8755. There's also `session.id` to handle, and we need to make sure `telemetry.sdk.elastic_export_timestamp` isn't recorded as a label. Assuming we'll do those as followups, please remove the "Closes" from the description.

kruskall commented:
> Can you please add tests which show that `host.os.{name,type}` are set correctly for iOS and Android?
Added 👍 

> Also, these changes aren't quite enough to close #8755. There's also `session.id` to handle, 

I see that the documentation doesn't mention a convention yet (see https://github.com/elastic/apm/blob/main/specs/agents/mobile/README.md#additional-span-attributes). Should we go with the RFC linked in the issue ?

> and we need to make sure `telemetry.sdk.elastic_export_timestamp` isn't recorded as a label.

I've pushed another commit to handle that and I've added assertions to verify that.

axw commented:
> I see that the documentation doesn't mention a convention yet (see https://github.com/elastic/apm/blob/main/specs/agents/mobile/README.md#additional-span-attributes). Should we go with the RFC linked in the issue ?

There's two parts to this that aren't yet specified:
 - the OTel side, there's no convention yet for encoding session IDs. However, the Elastic APM iOS agent is encoding this as a span attribute called `session.id`: https://github.com/elastic/apm-agent-ios/blob/93febce05fca0f407136b5066c9f87a79d44f86c/Sources/apm-agent-ios/Instrumentation/TapLogger.swift#L68
 - in ECS, there's only the RFC which you mentioned. Nevertheless, we've been recording `session.id` for RUM for a while already: https://github.com/elastic/apm-server/blob/b6ede69b407ee58ba3b9bbd4c8b63b970c28ab11/internal/model/apmevent.go#L50

So we should look for OTel `session.id` span attributes (for both transactions and spans I suppose), and record its value in the field `model.APMEvent.Session.ID`.

axw commented on internal/processor/otel/metadata.go:
```diff
@@ -106,6 +106,10 @@ func translateResourceMetadata(resource pcommon.Resource, out *model.APMEvent) {
 		case "host.arch":
 			out.Host.Architecture = truncate(v.StringVal())
 
+		// session.*
+		case "session.id":
+			out.Session.ID = truncate(v.StringVal())
```
This one will come from [span attributes](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/README.md), not [resource attributes](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/README.md).

In case you're not already familiar with the distinction:
 - resource attributes describe the entity producing observability data, and the environment in which that entity exists. This comprises things like `service.name`, `process.pid`, `os.name`, etc.
 - span attributes describe an individual operation, e.g. `http.method` for an HTTP request, or `db.statement` for a database query

Session IDs are event-specific, so we'll need to handle them in `traces.go`: I think both in `TranslateTransaction` and `TranslateSpan`.

kruskall replied:
Oh, I see. Thank you for the explanation! 🙇 

I've updated the code!

axw reviewed (approved):
LGTM, thanks :)
Please remember to update changelog(s).
